detach/detach.go

```diff
@@ -0,0 +1,97 @@
+package detach
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// Hooks are functions that are called with specific detach helpers to attach values from the parent context to the new
+// context. In order to be registered, at least one function in Hooks must be non-nil. Nil functions will be skipped.
+type Hooks struct {
+	// Inline functions are used to attach values from the src context to the dst context in calls to Inline.
+	//
+	// Inline functions should assume that the calling function will wait for the any calls after Inline to return before
+	// it returns, so they should just copy values from src to dst.
+	Inline func(dst, src context.Context) context.Context
+
+	// Async functions wrap the run methods passed to Async. They should copy values from the src context to the dst
+	// context and do any setup/cleanup around those values.
+	//
+	// Async functions should assume that the calling function could return before they finish, so if context values on
+	// src could be closed or otherwise cleaned up, the wrapper should likely create new values and clean those up rather
+	// than just copying the values from src to dst.
+	Async func(dst, src context.Context, next func(ctx context.Context))
+}
+
+type hooksRegistry struct {
+	mu    sync.RWMutex
+	hooks []Hooks
+}
+
+func (r *hooksRegistry) add(hooks Hooks) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.hooks = append(r.hooks, hooks)
+}
+
+func (r *hooksRegistry) allHooks() []Hooks {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.hooks
+}
+
+var registry = new(hooksRegistry)
+
+// Register adds a new Hooks to the library's registry. If all functions in the Hooks are nil, Register will panic.
+func Register(hooks Hooks) {
+	if hooks.Inline == nil && hooks.Async == nil {
+		panic("cannot register Hooks with all nil functions")
```

Comment:
🔕 since both `Inline` and `Async` would auto skip hooks without the corresponding field, this panic doesn't seem necessary, but I don't feel strongly either way.

Reply:
I agree that it's not strictly necessary, but I also think that handing in an empty Hooks is a bug or an oversight that we want to fix, which is why I added the panic.

Reply:
yeah agreed